Client/Frontend/Overlay/Content/Cheatsheet/BloomFilter/BloomFilterManager.swift

```
@@ -0,0 +1,308 @@
+                    return
+                }
+
+                guard let tempFilterURL = tempFilterURL else {
+                    completion(.failure(LoaderError.downloadedFileNotFound))
+                    return
+                }
+
+                do {
+                    guard let loadedFilter = try verifyAndLoad(from: tempFilterURL)
+                    else {
+                        throw LoaderError.invalidFilterBinaryFile
+                    }
+                    try? moveDownloadedFilter(from: tempFilterURL, to: locations.saveToURL)
+                    completion(.success(loadedFilter))
+                } catch {
+                    completion(.failure(error))
+                }
+            }
+            filterDownloadTask.resume()
+        }
+        task.resume()
+    }
+
+    static private func isLocalFileReadble(url: URL) -> Bool {
+        guard url.isFileURL,
+            fileManager.fileExists(atPath: url.path),
+            fileManager.isReadableFile(atPath: url.path)
+        else {
+            return false
+        }
+
+        return true
+    }
+
+    static private func verifyFile(at url: URL, with checksum: Checksum) throws -> Bool {
+        let fileHashes = try FileHasher.computeHashOfFile(at: url, using: [.sha256, .md5])
+        guard let shaHash = fileHashes[.sha256],
+            let md5Hash = fileHashes[.md5],
+            shaHash == checksum.sha,
+            md5Hash == checksum.md5
+        else {
+            return false
+        }
+        return true
+    }
+
+    static private func moveDownloadedFilter(from srcURL: URL, to dstURL: URL) throws {
+        if fileManager.fileExists(atPath: dstURL.path) {
+            try fileManager.removeItem(at: dstURL)
+        }
+
+        let dstDirURL = dstURL.deletingLastPathComponent()
+        if !fileManager.fileExists(atPath: dstDirURL.path) {
+            try createDirectory(dstDirURL)
+        }
+
+        try fileManager.moveItem(at: srcURL, to: dstURL)
```

If this file operation fails and we are then left with no bloom filter file, do we at least fail gracefully?

---

yes, we can `moveDownloadedFilter` with `try ?`, so it just fails silently. I changed it to catch and log